include/boost/math/special_functions/prime_sieve.hpp

```diff
@@ -0,0 +1,402 @@
+// Copyright 2020 Matt Borland
+//
+// Use, modification and distribution are subject to the
+// Boost Software License, Version 1.0.
+// (See accompanying file LICENSE_1_0.txt
+// or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#ifndef BOOST_MATH_SPECIAL_FUNCTIONS_PRIME_SIEVE_HPP
+#define BOOST_MATH_SPECIAL_FUNCTIONS_PRIME_SIEVE_HPP
+
+#include <boost/math/special_functions/prime.hpp>
+#include <boost/assert.hpp>
+#include <vector>
+#include <iterator>
+#include <cmath>
+#include <thread>
+#include <memory>
+#include <future>
+
+#ifdef _MSVC_LANG
+#if _MSVC_LANG >= 201703 // _MSVC_LANG == __cplusplus: https://devblogs.microsoft.com/cppblog/msvc-now-correctly-reports-__cplusplus/
+#include <execution>
+#endif
+#else
+#if __cplusplus >= 201703
+#include <execution>
+#endif
+#endif
+
+namespace boost { namespace math { namespace detail
+{
+// https://mathworld.wolfram.com/SieveofEratosthenes.html
+// https://www.cs.utexas.edu/users/misra/scannedPdf.dir/linearSieve.pdf
+template<class Z, class Container>
```

Comment:
I personally wouldn't use Z, since it is the name of an infinite _set_ of numbers; it doesn't really fit into a taxonomy of _concepts_. (And no computer integer type is infinite.) It fits into the taxonomy of _numbers_, in which there are various supersets of it such as R and Z[_i_], but only integer-like numbers (such as Z[_i_] but not R) will work here. Concepts can specify these properties, which is why it is important to use them as opposed to names of numbers.

So, the appropriate concept here is probably _DiscreteArchimedeanRing_, but no-one knows what that means unless they have memorized _Elements of Programming_. So everyone just writes _Integer_.  :) 
Now you might say, but that's what Z means! It's a subtle distinction, sure. "Z" is a set, but "Integer" is a concept.

And don't even get me started on R.

Reply:
With C++20 concepts would you use `template<Integer Z, class Container>`? I can see for now replacing Z with Integer making sense.

Reply:
I haven't spent much time thinking about how I would name template parameters using the concepts syntax in C++20, sorry.  :)